engine/src/main/java/org/terasology/engine/core/module/ClasspathCompromisingModuleFactory.java

```diff
@@ -52,6 +63,97 @@ public Module createArchiveModule(ModuleMetadata metadata, File archive) throws
                 new ClassesInModule(module));
     }
 
+    /**
+     * Find the location of the module containing this URL.
+     * <p>
+     * Accounts for loading modules from development workspaces that may have their build directories
+     * on the classpath, as is the case when running tests.
+     *
+     * @see #setDefaultCodeSubpath
+     * @see #setDefaultLibsSubpath
+     *
+     * @param metadataName the expected name of the metadata file, as it would appear in {@link #getModuleMetadataLoaderMap()}
+     * @param metadataUrl a URL of a metadata file, such as might be returned from {@link ClassLoader#getSystemResource}
+     * @return the module's base directory, or a jar file if it doesn't look like a local build
+     */
+    Path canonicalModuleLocation(String metadataName, URL metadataUrl) {
+        checkArgument(getModuleMetadataLoaderMap().containsKey(metadataName),
+                "metadataName `%s` is not in loader map", metadataName);
+        if (metadataUrl.getProtocol().equals("jar")) {
+            return modulePathFromMetadataJarUrl(metadataUrl);
+        } else {
+            return modulePathFromMetadataFileUrl(metadataName, metadataUrl);
+        }
+    }
+
+    private Path modulePathFromMetadataFileUrl(String metadata_name, URL url) {
+        Path path = fromUrl(url);
+        // We are considering the location of a resource file, so compare it to the code path.
+        // Include the metadata_name in case it has path components of its own.
+        Path relativePathFromModuleRoot = Paths.get(getDefaultCodeSubpath(), metadata_name);
+        return findModuleRoot(relativePathFromModuleRoot, path).orElse(path);
+    }
+
+    private Path modulePathFromMetadataJarUrl(URL url) {
+        checkArgument(url.getProtocol().equals("jar"), "Not a jar URL: %s", url);
+        try {
+            JarURLConnection connection = (JarURLConnection) url.openConnection();
+            url = connection.getJarFileURL();
```

Comment:
We're overriding the URL we get passed in here - is that a good practice? The URL we get passed in is not the same as the one we get from `getJarFileURL()`, or is it?

Reply:
It converts it from `jar:file:/stuff/Foo.jar!/module.txt` to `file:/stuff/Foo.jar`.

but oh, I moved the conditional out of here, that makes things easier. So if we want to keep a practice of treating method arguments as `final`, I can introduce another local variable for that.

Reply:
I'd prefer to treat method arguments as `final` (and even mark them as such). 👍